### PR TITLE
refact scion init

### PR DESCRIPTION
**Pull request by rakkit**
need merge https://github.com/janEbert/torchtitan/pull/26 first

So we provided two init functions from optimizer.distributed_scion to help init, no redundant code in upstream/component/optimizer.py

And we no longer take lots of configs to init. Instead, we can add them to a list

```
[[optimizer.extra_splits_rules]]
str_match = "tok_embeddings.weight"
norm_factor = "embed_sqrt"
backend = "identity"

[[optimizer.extra_splits_rules]]
str_match = "output.weight"
norm_factor = "unembed_sqrt"
backend = "identity"

[[optimizer.extra_splits_rules]]
str_match = "\\.router.gate.weight"
norm_factor = "spectral"

[[optimizer.extra_splits_rules]]
str_match = "ssnorm_scale"
norm_factor = "sign"
backend = "identity"
```

### CODE REVIEW

**Comment by rakkit** [comment]
force pushed after merging https://github.com/janEbert/torchtitan/pull/26

**Comment by rakkit** [comment]
rebased & force -pushed

File: torchtitan/components/optimizer.py [hunk 1]
```diff
@@ -44,55 +46,6 @@
...
-    param_names = list(param_dict.keys())
-    params.insert(
-        0,
-        {
-            "params": [param_dict.pop(n) for n in param_names],
-            "param_names": param_names,
-        },
-    )
-    assert not param_dict
-    return params
```
Comment (janEbert) [inline torchtitan/components/optimizer.py:93]: It's fine to move this function, but it should keep compatibility with and be available for the other supported optimizers.

File: torchtitan/components/optimizer.py [hunk 2]
```diff
@@ -131,37 +84,22 @@ def __init__(
...
-            extra_kwargs = kwargs.pop("extra_kwargs")
-            params = _extract_param_groups(model, kwargs)
-
-            is_scion = issubclass(optimizer_cls, (Scion, DistributedScion))
-            if is_scion:
-                kwargs.update(extra_kwargs)
-            self.optimizers.append(optimizer_cls(params, **kwargs))
+            else:
+                params = [p for p in model.parameters() if p.requires_grad]
+            self.optimizers.append(optimizer_cls(params, **optimizer_kwargs))
```
Comment (janEbert) [inline torchtitan/components/optimizer.py]: 1. As mentioned previously, this should keep compatibility with the other optimizers. What's the reason to remove the support for other optimizers? The `create_scion_param_groups` (previously `_extract_param_groups`) function didn't really change or have special Scion focus, except for the addition of `default_config`, which can be if-cased depending on the used optimizer.
2. `optimizer_kwargs` is being overwritten in the new implementation. Please use a new variable like before to prevent that (previously called `kwargs`).

File: torchtitan/optimizers/distributed_scion.py [hunk 3]
```diff
@@ -206,6 +208,12 @@ def get_norms_at_current_step(self):
         else:
             return {}
 
+    def zero_grad(self, *args, **kwargs):
+        if self.is_light:
+            pass
+        else:
+            super().zero_grad(*args, **kwargs)
```
Comment (janEbert) [inline torchtitan/optimizers/distributed_scion.py]: Can you add an extra flag `force=False` to zero grads even for the light version?

File: torchtitan/components/loss.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ def multi_token_cross_entropy_loss(
     main_loss = loss_fn(preds[0], labels[:, : job_config.training.seq_len])
 
     mtp_loss = 0
-    for (label_offset, pred) in enumerate(preds[1:], 1):
+    for label_offset, pred in enumerate(preds[1:], 1):
```
Comment (janEbert) [inline torchtitan/components/loss.py]: Can you maybe configure your auto formatter not to change these things? Let's just take the TorchTitan pre-commit hooks as the ground truth.

File: torchtitan/components/optimizer.py [hunk 5]
```diff
@@ -474,101 +399,16 @@ def build_optimizers(
...
-
-    optimizer_kwargs["extra_kwargs"] = {
-        "parallel_dims": parallel_dims,
-        **extra_kwargs,
-    }
-
     optimizer_classes = {
         "Adam": torch.optim.Adam,
         "AdamW": torch.optim.AdamW,
-        "Scion": Scion,
```
Comment (janEbert) [inline torchtitan/components/optimizer.py:571]: We should probably also remove its files if we officially remove it. Could you also explain how you ensured equality between the distributed and non-distributed version?

File: torchtitan/config/manager.py [hunk 6]
```diff
@@ -23,6 +25,70 @@
...
+    remaining, overrides = [], []
+    for a in raw_args:
+        if a.startswith("-") and "[" in a and "=" in a:
+            key, val = a.lstrip("-").split("=", 1)
+            key = key.replace("-", "_")  # normalise hyphens to underscores
+            path = key.split(".")
+            overrides.append((path, _literal_eval_safe(val)))
+        else:
+            remaining.append(a)
+    return remaining, overrides
```
Comment (janEbert) [inline torchtitan/config/manager.py]: 1. Could this be implemented with JSON/TOML parsing instead of `literal_eval`? This feels hacky with how it manually tries to find things with the square brackets and equals literal search, then parsing Python code.
2. What do we need this for?

File: torchtitan/datasets/hf_datasets.py [hunk 7]
```diff
@@ -649,6 +649,17 @@ def build_hf_validation_dataloader(
...
+        dataset=hf_ds,
+        seq_len=seq_len,
+        infinite=False,
+        dataset_inner_name=dataset_inner_name,
+        dataset_files=dataset_files,
+        dataset_split=dataset_split,
+        dataset_streaming=dataset_streaming,
+        dataset_key=dataset_key,
+    )
+
```
Comment (janEbert) [inline torchtitan/datasets/hf_datasets.py]: Please remove, this is broken.

File: torchtitan/models/llama3/model/model.py [hunk 8]
```diff
@@ -736,6 +736,9 @@ def forward(
                 This will always be the input batch regardless of the pipeline stage.
                 This field is required for non-first PP stages to perform document
                 masking attention (to analyze the boundary of the document).
+            prev_embed (torch.Tensor | None): Output token embeddings of
+                previous Transformer layer (after output norm, before
+                unembedding).
```
Comment (janEbert) [inline torchtitan/models/llama3/model/model.py]: Please remove, duplicate.

File: torchtitan/config/job_config.py [hunk 9]
```diff
@@ -260,20 +263,8 @@ class Optimizer:
...
-
-    embed_str_match: str | None = None
-    """String to match for embedding layer parameter group"""
-
-    unembed_str_match: str | None = None
-    """String to match for unembedding layer parameter group"""
-
-    router_str_match: str | None = None
-    """String to match for MoE router layer parameter group"""
+    extra_splits_rules: list[dict[str, Any]] | None = None
```
Comment (janEbert) [inline torchtitan/config/job_config.py]: Can we rename it to something more indicative? `[extra_]param_group_[split_]rules`?

File: torchtitan/optimizers/utils.py [hunk 10]
```diff
@@ -18,3 +24,183 @@ def remove_orig_mod_and_weight_for_p_name(name: str) -> str:
...
+            "param_names": param_names,
+        }
+        # Add default configuration to the default group
+        default_group.update(default_config)
+        params.insert(0, default_group)
+
+    # Create clean kwargs for the optimizer (remove extra_splits_rules)
+    clean_kwargs = {
+        k: v for k, v in optimizer_kwargs.items() if k != "extra_splits_rules"
+    }
```
Comment (janEbert) [inline torchtitan/optimizers/utils.py]: Maybe just `copy`+`del`?

**Review by janEbert — CHANGES_REQUESTED** [review]
The refactoring of optimizer settings is nice, but some things have been lost along the way. Maybe there are reasons for that which I don't understand, so feel free to explain.

File: torchtitan/components/optimizer.py (hunk 5 above)
Comment (rakkit) [reply torchtitan/components/optimizer.py:571]: we should remove it, we cannot ensure the "equality" because its not well tested for quiet long

File: torchtitan/config/manager.py (hunk 6 above)
Comment (rakkit) [reply torchtitan/config/manager.py]: we need it for "extra_splits_rules" or "[extra_]param_group_[split_]rules" to set scion's configs via cli.
e.g. 
```
--optimizer.extra_splits_rules[0].lr=$lr \
--optimizer.extra_splits_rules[1].lr=$lr \
```

File: torchtitan/components/optimizer.py (hunk 1 above)
Comment (rakkit) [reply torchtitan/components/optimizer.py:93]: ok, but if we dont have group-wise configs, we can merge this function with scion.
e.g. this function cannot set separately lr or wd for embedding/output in AdamW. Then it just split the params into groups, but all groups have the same configs.

but its a good point, the refactor "create_scion_param_groups" can be intend to make it general allows us to do  group-wise configs, for all optimizers

File: torchtitan/components/optimizer.py (hunk 5 above)
Comment (janEbert) [reply torchtitan/components/optimizer.py:571]: So the test was a loss curve comparison? Maybe checking gradients/weights would make sense to gain more confidence.

File: torchtitan/config/manager.py (hunk 6 above)
Comment (janEbert) [reply torchtitan/config/manager.py]: Got it; so this assumes a lot of things and can break super easily if the TOML changes. I wonder if there is another way.

File: torchtitan/optimizers/distributed_scion.py (hunk 3 above)
Comment (rakkit) [reply torchtitan/optimizers/distributed_scion.py]: What does `force=False ` mean?

**Comment by rakkit** [comment]
fixed nvshmem and disable it by default. (nvshmem disco only work on single node)
